---
title: "Useful R functions"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(magrittr) # for: %>%
library(ape) # for: read.tree, plot.phylo
library(seqinr) # for: read.fasta, write.fasta
```

## Reading and plotting trees
```{r,echo=F}
# Read in .treefile generated by IQ-TREE
tree <- read.tree('BHV1_plus BHV5 outgroup_alignment.txt.treefile')

# Plot the phylogeny and save image
pdf("BHV1-plus-BHV5-outgroup-alignment-notscaled-cladogram.pdf",
    width=12,height=10)
# There are range of options for `type`
# Do not use edge lengths to better visualize the topology
plot.phylo(tree,type="cladogram",use.edge.length=F)
dev.off()
```

## Reading, subsetting & writing FASTA file
```{r}
# FASTA file is read in as a list, where the entry names are the taxa names
# and each entry stores a character vector which corresponds to the aligned
# sequence for a given taxa
full.fasta <- read.fasta(file="BHV1_plus BHV5 outgroup_alignment.txt")
# length(full.fasta) # 50 taxa
# full.fasta$BHV5 %>% length # 144551 sites

# strains for subset analysis
subset <- c("C14_CSU_034_10640","C33","MN5","MN12","MN3","C46",
            "BoviShield_Gold_FP5_MLV_vaccine")
subset.fasta <- full.fasta[match(subset,names(full.fasta))]
# length(subset.fasta) # 7 taxa
# names(subset.fasta)
# subset.fasta$C14_CSU_034_10640 %>% length # 144551 sites

write.fasta(sequences=subset.fasta,names=names(subset.fasta),
            file.out="BHV1_plus BHV5 outgroup_alignment subset.fasta")
```



